include/revng/Support/MetaAddress/IntervalContainers.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstdint>
+
+#include "boost/icl/interval_map.hpp"
+
+#include "revng/Support/MetaAddress.h"
+
+namespace boost::icl {
+
+template<>
+struct is_discrete<MetaAddress> {
+  using type = is_discrete;
+  BOOST_STATIC_CONSTANT(bool, value = true);
+};
+
+template<>
+struct identity_element<MetaAddress> {
+  static MetaAddress value() { return MetaAddress{}; }
+};
+
+template<>
+struct has_difference<MetaAddress> {
+  using type = has_difference;
+  BOOST_STATIC_CONSTANT(bool, value = true);
+};
+
+template<>
+struct difference_type_of<MetaAddress> {
+  // assume we compare only .address() within same type, epoch and address space
+  using type = uint64_t;
```

Comment:
Why is it returning `std::optional<uint64_t>`?

Reply:
But it looks like for `MetaAddress.Address` always have some value. It's primitive type. So `operator-(MetaAddress)` will always have some value, because [here](https://github.com/revng/revng/blob/aa74e194bcdac0849fa47c4f9c4d8524c0781369/include/revng/Support/MetaAddress.h#L646C23-L646C23) we use `OverflowSafeInt` and it's `operator-` ([here](https://github.com/revng/revng/blob/aa74e194bcdac0849fa47c4f9c4d8524c0781369/include/revng/Support/OverflowSafeInt.h#L42) will always return optional with value). We construct `OverflowSafeInt` with primitive value, not optional, so it's `Value` is set for both addresses (MetaAddresses), so the result should be safely unwrapped from optional. Can't we add an assertion on `has_value` there and return bare `uint64_t`?

Reply:
If you do `MetaAddress(100) - MetaAddress(101)` it should return a nullopt.

Reply:
Correct, I see now that `OverflowSafeInt::sub` can return `std::nullopt`. Could you suggest any solution? Should I use some other key type for interval_map?

Reply:
@kolodziej is this fixed?

Reply:
Done